### DISCUSSION TITLE
Can you create entities in multi-threaded systems? Are observers multi-threaded?

### DISCUSSION THREAD

**Discussion by destructor465** [SanderMertens/flecs#1198] [Q&A] [answered]
So I been trying to use flecs in my C++ project. And let me just say it's awesome. I have a few questions if somebody could help me to answer them:
1. Can you create entities in multi-threaded systems? So far I been trying to call `world.entity().add<SomeComponent>();` in the multi-threaded system but it results in fatal error `flecs.c: 6438: assert: ecs_is_alive(world, entity) INVALID_PARAMETER`. I also tried deferring this operation to no luck. So I wonder is it possible to create entities in multi-threaded systems?
2. Do observers utilize multi-threading or do they always run on the same thread where you call `progress`?

**Comment by SanderMertens** [comment, marked as answer]
1. You _can_ create entities from a multithreaded system but it's very limited what you can do with those, because they can't be made alive yet while the world is in "readonly" mode. One workaround for that is to create a pool of alive entity ids beforehand.
2. Observers don't use multithreading.

**Reply by destructor465** [reply]
Perfect, thank you so much.

**Reply by Jernesstar** [reply]
Hello, I'm facing a similar issue and would like to know how to go about making this pool of alive entity ids

**Reply by SanderMertens** [reply]
Could be as simple as having a `vector<flecs::entity> unused_entities`, and have threads consume elements from that vector if they want to create an entity.
